Add breeze switch to Renson integration

jimmyd-be opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Added breeze switch to manually switch to breeze mode of the ventilcation unit.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/31462

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jimmyd-be committed:
Add breeze switch

joostlek commented on homeassistant/components/renson/sensor.py:
```diff
@@ -184,14 +184,14 @@ class RensonSensorEntityDescription(
         device_class=SensorDeviceClass.TEMPERATURE,
         native_unit_of_measurement=UnitOfTemperature.CELSIUS,
-        entity_registry_enabled_default=False,
+        entity_registry_enabled_default=True,
```
```suggestion
```
Default is True

joostlek commented on homeassistant/components/renson/sensor.py:
```diff
@@ -184,14 +184,14 @@ class RensonSensorEntityDescription(
         field=BREEZE_LEVEL_FIELD,
         raw_format=False,
-        entity_registry_enabled_default=False,
+        entity_registry_enabled_default=True,
```
```suggestion
```

joostlek commented on homeassistant/components/renson/switch.py:
```diff
@@ -0,0 +1,91 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class RensonBreezeSwitch(RensonEntity, SwitchEntity):
```
Can you maybe explain what the difference is between ventilation and breeze?

jimmyd-be replied:
Breeze is a functionality of the device that pushes the outside air into the home
Ventilation then their will be heat exchangers between the air outside and inside to ventilate the house. Ventilation will not cooldown the house but filters the air inside the house.

joostlek replied:
Ah right, how does Breeze work with the different speed levels? Since you also have the fan mode breeze and you have several levels.

jimmyd-be replied:
Fan cannot work at the same time as Breeze function. The fan can set level of fan speed directly. For Breeze it is needed to change the Breeze configuration to change the levels.

joostlek replied:
So if you can set the unit to Breeze via the fan entity, why do we need a separate switch?

(just trying to understand the why behind this change)

jimmyd-be replied:
We need a way to tell the device to use the breeze instead of the ventilation. The only way that I can think of to use only the fan for both ventilation and breeze if it is possible to set the speed level with negative values. Then we can use the negative values to the breeze and positive functions to ventilation (but don't think this is posible). Therefore this switch to enable the breeze function and use the ventilation level to configure the breeze level instead of the venitilcation level.

jimmyd-be replied:
Or we can create two fan sensors, one for ventilation and one for breeze. Than can also work

joostlek replied:
Right, so if I understand correctly, you want to use the ventilation levels of the fan to control the level of breeze/ventilation based on the value of the switch?

Did I understand correctly?

Are there more modes than breeze/ventilation?

jimmyd-be replied:
That's correct. No there are no more modes that I know of.

joostlek replied:
I think I got the picture now.

The thing I am doubting about is that a switch is usually for toggling stuff on and off. We don't really use them for mode A - mode B. I think in general we use select entities for this (there probably are cases in the code where we do use the switch, but that might be before the select was introduced).

I think merging both parts into one fan isn't ideal.

Two fans might work. as long as they can update correctly. So if I select level 3 on breeze, the other fan turns off.

I do still doubt if that is a correct way. So don't see this a definitive answer on what you should do, but more as a food for thought.

jimmyd-be replied:
I implented the fan approche.

joostlek commented on homeassistant/components/renson/switch.py:
```diff
@@ -0,0 +1,91 @@
+    """Provide the breeze switch."""
+
+    _attr_icon = "mdi:weather-dust"
+    _attr_name = "Breeze"
```
can we use translation keys and `has_entity_name = True`?

joostlek commented on homeassistant/components/renson/switch.py:
```diff
@@ -0,0 +1,91 @@
+
+        self._state = None
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if switch is on."""
+        if self._state is None:
+            return False
+
+        return self._state
```
Please use `_attr_is_on` instead of `_state`, that way you can remove this property

joostlek commented on homeassistant/components/renson/switch.py:
```diff
@@ -0,0 +1,91 @@
+
+        return self._state
+
+    @property
+    def device_class(self) -> SwitchDeviceClass:
+        """Set renson switch device class."""
+        return SwitchDeviceClass.SWITCH
```
can be a shorthand attribute

jimmyd-be committed:
Cleanup code

joostlek commented on homeassistant/components/renson/switch.py:
```diff
@@ -0,0 +1,80 @@
+        """Initialize class."""
+        super().__init__("breeze", api, coordinator)
+
+        self._attr_is_on = False
```
You can move this outside of the constructor and remove the constructor completely

jimmyd-be committed:
Replace switch entity with new fan entity

emontnemery commented:
@jimmyd-be Please try to better explain what this PR does. Is it in reality a single renson device, which you want to model as two fan entities?

krmarien commented:
I added on the original version of this component (when it was not yet home-assistant/core) the breeze feature with a switch. https://github.com/krmarien/Home-Assistant-Sensor-Renson-Ventilation/tree/breeze

To me that makes more sense as this is still one physical device (fan), it will just toggle some valves if I'm right.

emontnemery commented:
I'm marking this PR as draft since there are open questions about the implementation.
@jimmyd-be Please reply to the question, then click the "Ready for review" button 👍

jimmyd-be commented:
This thread contains the most info https://github.com/home-assistant/core/pull/101641#discussion_r1349657335

jimmyd-be commented:
> I added on the original version of this component (when it was not yet home-assistant/core) the breeze feature with a switch. https://github.com/krmarien/Home-Assistant-Sensor-Renson-Ventilation/tree/breeze
> 
> To me that makes more sense as this is still one physical device (fan), it will just toggle some valves if I'm right.

I first used your implementation but after first review it was not OK for the comunity. And changed after that to the current implementation.

jimmyd-be commented:
> @jimmyd-be Please try to better explain what this PR does. Is it in reality a single renson device, which you want to model as two fan entities? If that's the case, why is that?

It is a single device. And the device can push air from outside to inside and visa versa. The breeze function is for pushing outside air into the home for cooling. The normal fan is for filtering the air and with a temperature value to retain the temperature inside the house.

emontnemery commented:
sorry @jimmyd-be, I didn't realize the recommendation to model the breeze mode as an additional fan came from review comments.

jimmyd-be commented:
> sorry @jimmyd-be, I didn't realize the recommendation to model the breeze mode as an additional fan came from review comments.

No problem, what do I need to change now? Keep this implementation or revert back to the switch and fan implementation?

emontnemery commented:
@jimmyd-be I'll check with @joostlek and a few others.

joostlek commented:
Just had a quick chat with Erik, and I want to apologise for the wrong recommendation. Back then I didn't know the things I know today.

I think having 1 fan entity like we already had and then have a switch or select to switch the mode would be the best.

jimmyd-be committed:
Revert "Replace switch entity with new fan entity"

This reverts commit 4fc1ac22e67091a7e980aefd217652f2a88bed17.

emontnemery commented:
@jimmyd-be please update the title and the description of the PR to match the implementation

jimmyd-be commented:
> @jimmyd-be please update the title and the description of the PR to match the implementation

Update the PR title and description. Reverted the code back to the previous commit.

emontnemery commented on homeassistant/components/renson/sensor.py:
```diff
@@ -172,14 +172,12 @@ class RensonSensorEntityDescription(
         translation_key="breeze_level",
         field=BREEZE_LEVEL_FIELD,
         raw_format=False,
-        entity_registry_enabled_default=False,
```
Why is this sensor no longer disabled, isn't it enough to show the level as the fan speed?

jimmyd-be replied:
Breeze level is a setting inside the device and can be set using the official application. When toggling the breeze switch it will take this value as fan speed.

emontnemery replied:
Yes, i saw it's used in the fan entity to set and get fan speed when breeze mode is enabled. Why is an additional sensor useful?

jimmyd-be replied:
More info about the device that users can view.

emontnemery replied:
OK

emontnemery reviewed (commented):
Looks good to me, just one question.

emontnemery reviewed (approved):
LGTM, thanks @jimmyd-be 👍
